pkg/storegateway/indexheader/encoding/factory.go

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package encoding
+
+import (
+	"bufio"
+	"encoding/binary"
+	"hash/crc32"
+	"os"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+const bufferSize = 4096
+
+// DecbufFactory creates new file-backed decoding buffer instances for a specific index-header file.
+type DecbufFactory struct {
+	pool sync.Pool
+	path string
+}
+
+func NewDecbufFactory(path string) *DecbufFactory {
+	return &DecbufFactory{
+		pool: sync.Pool{
+			New: func() any {
+				return bufio.NewReaderSize(nil, bufferSize)
+			},
+		},
+		path: path,
+	}
+}
+
+// NewDecbufAtChecked returns a new file-backed decoding buffer positioned at offset + 4 bytes.
+// It expects the first 4 bytes after offset to hold the big endian encoded content length, followed
+// by the contents and the expected checksum. This method checks the CRC of the content and will
+// return an error Decbuf if it does not match the expected CRC.
+func (df *DecbufFactory) NewDecbufAtChecked(offset int, table *crc32.Table) Decbuf {
+	f, err := os.Open(df.path)
```

Comment:
[non blocking] My biggest concern is that we open a file description each time. It's called in called in hot path, like `Symbols.Lookup()` and `Symbols.ReverseLookup()`. Not a blocker for this PR, but I would like to hear your thoughts on this.

Reply:
I've been worrying about this but after deploying a test image in a dev cluster, opening the file doesn't seem to account for much CPU time (according to phlare). The zone with this index-header reader enabled seems to be about as fast as the existing mmap implementation as far as I can tell.

If it turned out to be an issue, I was thinking about maintaining a (small) pool of file handles for each index-header.

Reply:
Actually, maybe it is significant. Looking more closely at the profile: `Symbols.Lookup()` is about 0.27% of total time and `os.OpenFile` is 0.12% of total time (based on the load test that we've been running). So I guess 1/2 the time spent by the symbol reader is opening the file even if it's not a large part of time spent overall. I'll add pooling of file handles to the list of future performance improvements.

Reply:
>  I'll add pooling of file handles to the list of future performance improvements.

👍